File: adventofcode/2024/04_2.cpp
#include <bits/stdc++.h>
#define int long long

using namespace std;
vector<string> grid;

int rows, cols;
vector<vector<pair<int, int>>> dirs = {
    {{-1, -1}, {0, 0}, {1, 1}},
    {{-1, 1}, {0, 0}, {1, -1}},
};
const string MAS = "MAS";
const string SAM = "SAM";

int search(int i, int j) {
    int ans = 0;
    for(auto d: dirs) {
        string s = "";
        for(int k=0; k<3; k++) {
            auto [dx, dy] = d[k];
            int x=i+dx;
            int y=j+dy;
            if (x<0 || x>=rows || y<0 || y>=cols) return 0;
            s += grid[x][y];
        }
        if (s != MAS && s != SAM) return 0;
    }
    return 1;
}

signed main() {
    ios::sync_with_stdio(0);
    cin.tie(0);
    #ifndef ONLINE_JUDGE 
        freopen("../input.txt", "r", stdin); 
        freopen("../output.txt", "w", stdout); 
    #endif
    const int LINES=140;

    for(int i=0; i<LINES; i++)
    {
        string s; getline(cin, s);
        grid.push_back(s);
    }
    rows = grid.size();
    cols = grid[0].size();
    int ans = 0;
    for(int i=0; i<rows; i++) {
        for(int j=0; j<cols; j++) {
            ans += search(i, j);
        }
    }
    cout << ans;
    return 0;
}

File: adventofcode/2024/python/04_2.py
import sys

dirs = [
    [(-1, -1), (0, 0), (1, 1)],
    [(-1, 1), (0, 0), (1, -1)],
]

MAS = "MAS"
SAM = "SAM"
rows, cols = 0, 0
grid = []

def search(i, j):
    ans = 0
    for d in dirs:
        s = ""
        for k in range(3):
            dx, dy = d[k]
            x, y = i+dx, j+dy
            if x<0 or x>=rows or y<0 or y>=cols:
                return 0
            s += grid[x][y]
        if s != MAS and s != SAM:
            return 0
    return 1

def main():
    sys.stdin = open("../input.txt", "r")
    sys.stdout = open("../output.txt", "w")
    
    global rows, cols, grid
    LINES = 140
    grid = [input().strip() for _ in range(LINES)]
    rows, cols = len(grid), len(grid[0])
    ans = 0
    
    for i in range(rows):
        for j in range(cols):
            ans += search(i, j)
    
    print(ans)

if __name__ == "__main__":
    main()